### PR TITLE
CC Libraries - apply char style / layer style / color / color theme

### PR DESCRIPTION
**Pull request by shaoshing**
Now you can apply char style / layer style / color / color theme to the selected layers.

This also addresses issue #1983

### CODE REVIEW

File: src/js/actions/libraries.js [hunk 1]
```diff
@@ -514,31 +520,34 @@ define(function (require, exports) {
...
 
         var representation = element.getPrimaryRepresentation(),
             styleData = representation.getValue("characterstyle", "data"),
-            layerRef = textLayerAdapter.referenceBy.current,
+            textLayerIDs = collection.pluck(textLayers, "id"),
+            layerRef = textLayerIDs.map(textLayerAdapter.referenceBy.id).toArray(),
             applyObj = textLayerAdapter.applyTextStyle(layerRef, styleData);
 
-        return descriptor.playObject(applyObj)
+        return layerActionsUtil.playSimpleLayerActions(currentDocument, textLayers, applyObj, true)
```
Comment (shaoshing) [inline src/js/actions/libraries.js]: I have to make the changes here to only apply char style to text layers to address an issue (Set command is not available) when you select all layers (including the art board).

File: src/js/actions/type.js [hunk 2]
```diff
@@ -264,7 +265,7 @@ define(function (require, exports) {
             playObject = [playObject].concat(setOpacityPlayObjects);
         }
         var updatePromise = this.transfer(updateColor, document, layers, color, coalesce, true),
-            setColorPromise = locking.playWithLockOverride(document, layers, playObject, typeOptions);
+            setColorPromise = layerActionsUtil.playSimpleLayerActions(document, layers, playObject, true, typeOptions);
```
Comment (shaoshing) [inline src/js/actions/type.js]: This also address the `Set command is not available` issue when apply font color to all layers (including the art board)

**Comment by shaoshing** [comment]
@volfied please review.

File: src/js/actions/libraries.js [hunk 3]
```diff
@@ -514,31 +520,34 @@ define(function (require, exports) {
...
-     * @param {AdobeLibraryElement} element [description]
+     * @param {AdobeLibraryElement} element
      *
      * @return {Promise}
      */
     var applyCharacterStyle = function (element) {
         var appStore = this.flux.store("application"),
-            currentDocument = appStore.getCurrentDocument();
+            currentDocument = appStore.getCurrentDocument(),
+            selectedLayers = currentDocument ? currentDocument.layers.selected : new Immutable.List(),
```
Comment (volfied) [inline src/js/actions/libraries.js]: you don't need `new` here.

File: src/js/actions/libraries.js [hunk 4]
```diff
@@ -546,6 +555,50 @@ define(function (require, exports) {
...
+                return textLayers.isEmpty() ? null :
+                    this.transfer(typeActions.setColor, currentDocument, textLayers, color, true, true);
+            })
+            .then(function () {
+                return vectorLayers.isEmpty() ? null :
+                    this.transfer(shapeActions.setFillColor, currentDocument, vectorLayers, 0, color,
+                        true, true, true);
+            })
+            .then(function () {
+                return this.transfer(layerActions.resetLayers, currentDocument, textLayers.concat(vectorLayers));
```
Comment (volfied) [inline src/js/actions/libraries.js]: Do you need to resetLayers for color changes? Make sure it's absolutely necessary

File: src/js/actions/libraries.js [hunk 5]
```diff
@@ -546,6 +555,50 @@ define(function (require, exports) {
...
+            vectorLayers = selectedLayers.filter(function (l) { return l.isVector(); });
+
+        if (textLayers.isEmpty() && vectorLayers.isEmpty()) {
+            return Promise.resolve();
+        }
+
+        // FIXME: Setting font color and fill color at the same time will result in two histroy states.
+        //        We should merge the two states when transaction is supported.
+        return Promise.bind(this)
+            .then(function () {
```
Comment (volfied) [inline src/js/actions/libraries.js]: You should use Promise.join instead, look at actions/sampler.js/pasteStyle for an example.

File: src/js/actions/type.js (hunk 2 above)
Comment (volfied) [inline src/js/actions/type.js]: Given that you've fixed applyColor in libraries, and you only pass text layers to this function, this change should be unnecessary.

Comment (volfied) [inline src/js/actions/type.js]: Also, the only reason playSimpleLayerActions worked but playWithLockOverride didn't is latter skips errors.

File: src/js/jsx/sections/libraries/assets/AssetSection.jsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+         *
+         * @private
+         */
+        _listenTitleEvent: function () {
+            if (!this.refs.assetTitle) {
+                return;
+            }
+            
+            var assetTitleNode = React.findDOMNode(this.refs.assetTitle);
+            assetTitleNode.addEventListener("dblclick", this._handleRename);
```
Comment (volfied) [inline src/js/jsx/sections/libraries/assets/AssetSection.jsx]: Is there a reason for not using React's own `onDoubleClick` handler?

File: src/js/jsx/sections/libraries/assets/AssetSection.jsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        },
+
+        /**
+         * Handle select asset event.
+         * @param  {SynthenticEvent} event
+         *
+         * @private
+         */
+        _handleSelect: function (event) {
+            if (event.target !== React.findDOMNode(this)) {
```
Comment (volfied) [inline src/js/jsx/sections/libraries/assets/AssetSection.jsx]: This seems weird as well.. clickEvent should only come here if this component is indeed the target. If you're getting click events from one of it's children, you probably want to stopPropagation in the children's click handlers.

**Comment by volfied** [comment]
Looks good, but I have a lot of reservations about the implementation of AssetSection. Back to you @shaoshing

File: src/js/actions/type.js (hunk 2 above)
Comment (shaoshing) [inline src/js/actions/type.js]: The previous use of `playSimpleLayerActions` is for `applyCharacterStyle`, so it is necessary here to make `applyColor` works. I checked how `playSimpleLayerActions` works and looks like it will first select the desired layers for "Set" command, and then restore the previous select; this fix the issue when all layers are selected while apply fond color.

File: src/js/actions/libraries.js (hunk 4 above)
Comment (shaoshing) [inline src/js/actions/libraries.js]: Confirmed that both fills#setColor and shapes#setFillColor will dispatch the color change event, so we don't need the resetLayers action here.

File: src/js/jsx/sections/libraries/assets/AssetSection.jsx (hunk 6 above)
Comment (shaoshing) [inline src/js/jsx/sections/libraries/assets/AssetSection.jsx]: assetTitle is a reference of its children `this.props.children[0]`, and the regular React way of binding event is not available here. However, I figured out I can set the `onDoubleClick` option in the `React.cloneElement`. A optimized version will arrive soon

File: src/js/jsx/sections/libraries/assets/AssetSection.jsx (hunk 7 above)
Comment (shaoshing) [inline src/js/jsx/sections/libraries/assets/AssetSection.jsx]: The reason is that we are trying to support both click (select) and double click (rename / edit), and a double click event will also trigger a click event. I agree with you that this is not the optimal way, and I will find a better solution when I am working on the rename and edit feature.

**Comment by shaoshing** [comment]
hey @volfied , I made some improvements (thanks for the feedback), along with two additional fixes:

* missing collaborator icon
* missing add-text-color button in the contextual bar.

Please review again.

**Comment by shaoshing** [comment]
Resolved a bug of setting color for unexpanded group(s). https://github.com/adobe-photoshop/spaces-design/pull/2033/files#diff-d036559733279e3c642c1fffe3d0fbc8R568

File: src/js/actions/libraries.js [hunk 8]
```diff
@@ -157,6 +162,7 @@ define(function (require, exports) {
             .bind(this)
             .then(function (tempFilename) {
                 // Export the selected layers
+
```
Comment (volfied) [inline src/js/actions/libraries.js]: :space_invader: (unnecessary newline)

File: src/js/actions/libraries.js [hunk 9]
```diff
@@ -41,9 +42,13 @@ define(function (require, exports) {
         locks = require("../locks"),
         layerActions = require("./layers"),
         searchActions = require("./search/libraries"),
+        shapeActions = require("./shapes"),
```
Comment (volfied) [inline src/js/actions/libraries.js]: While you're in here, can you also sort this require list to make it a bit more readable.

File: src/js/actions/libraries.js [hunk 10]
```diff
@@ -546,6 +555,39 @@ define(function (require, exports) {
...
+     *  - Text layer: will set layer's font color
+     *  - Vector layer: will set layer's fill color
+     *
+     * @param {Color} color
+     *
+     * @return {Promise}
+     */
+    var applyColor = function (color) {
+        var currentDocument = this.flux.store("application").getCurrentDocument(),
+            selectedLayers = currentDocument ? currentDocument.layers.allSelected : null;
```
Comment (volfied) [inline src/js/actions/libraries.js]: You probably want to set it to Immutable.List() as well like above.

File: src/js/jsx/sections/libraries/LibraryBar.jsx [hunk 11]
```diff
@@ -165,7 +170,22 @@ define(function (require, exports, module) {
 
         render: function () {
             var adobeStockURL = "https://stock.adobe.com";
-            
+
+            var addFillButton = this._createColorButton(function (layer) {
+                return layer.fills.isEmpty() ? null :
```
Comment (volfied) [inline src/js/jsx/sections/libraries/LibraryBar.jsx]: Do we want these to be null, or disabled? Is that going to be in a separate PR?

File: src/js/jsx/sections/libraries/assets/AssetSection.jsx [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        },
+
+        /**
+         * Handle rename asset event.
+         * @param  {SynthenticEvent} event
+         *
+         * @private
+         */
+        _handleRename: function (event) {
+            event.stopPropagation();
```
Comment (volfied) [inline src/js/jsx/sections/libraries/assets/AssetSection.jsx]: Can you add a `// FIXME: Implement` here?

File: src/js/jsx/sections/libraries/assets/CharacterStyle.jsx [hunk 13]
```diff
@@ -80,40 +74,41 @@ define(function (require, exports, module) {
                 fontSize = charStyle.fontSize,
```
Comment (volfied) [inline src/js/jsx/sections/libraries/assets/CharacterStyle.jsx]: Probably a safe bet to just return `null` if charStyle.adbeFont is null.

File: src/js/jsx/sections/libraries/assets/ColorTheme.jsx [hunk 14]
```diff
@@ -29,56 +29,57 @@ define(function (require, exports, module) {
...
-                
-            var colorSwatchComponents = colorSwatches.map(function (color) {
-                var colorHex = tinycolor(color[0].value).toHexString().toUpperCase();
-                return (<div key={colorHex} style={{ background: colorHex }} title={colorHex}/>);
+
+            var colorSwatchComponents = colorSwatches.map(function (colors) {
+                var colorData = colors[0],
+                    colorHex = tinycolor(colorData.value).toHexString().toUpperCase();
+
+                return (<div key={colorHex}
```
Comment (volfied) [inline src/js/jsx/sections/libraries/assets/ColorTheme.jsx]: If the color theme has the same colors, this is going to have identical keys and React will be mad at us. Maybe use the `index` coming from the map function here instead?

**Comment by volfied** [comment]
One more round, looks great! @shaoshing

File: src/js/jsx/sections/libraries/LibraryBar.jsx (hunk 11 above)
Comment (shaoshing) [inline src/js/jsx/sections/libraries/LibraryBar.jsx]: It will be null here, same as standard PS. But we can definitely have a separate PR if we decide to show disabled button instead.

File: src/js/jsx/sections/libraries/assets/ColorTheme.jsx (hunk 14 above)
Comment (shaoshing) [inline src/js/jsx/sections/libraries/assets/ColorTheme.jsx]: Good catch!

**Comment by shaoshing** [comment]
Thanks for the review @volfied . I added more changes to address your comments.

**Comment by volfied** [comment]
You forgot two nits... but they're minor enough to not stop progress
